### PR TITLE
show hint at the bottom of link out of frame

### PR DESCRIPTION
**Pull request by wentaozou1**
I have fix [issue #3810](https://github.com/qutebrowser/qutebrowser/issues/3810). Now, if a clickable images that start out of frame will have their hints displayed at the left bottom corner. But still, if the picture shows less than 3px in the frame, the hint will not be displayed because there is not enough room for the label.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3827)

### CODE REVIEW

**Comment by mschilli87** [comment]
> But still, if the picture shows less than 3px in the frame, the hint will not be displayed because there is not enough room for the label.

Is that different for pictures showing less than 3px on the bottom? If yes, that might help to solve the 'problem'. Otherwise, it's consistent, so why bother? :wink:

File: qutebrowser/browser/hints.py [hunk 1]
```diff
@@ -124,7 +124,10 @@ def _move_to_elem(self):
             return
         no_js = config.val.hints.find_implementation != 'javascript'
         rect = self.elem.rect_on_view(no_js=no_js)
-        self.move(rect.x(), rect.y())
+        if rect.y()<0 and (rect.y()+rect.height()-15>0):
```
Comment (jgkamat) [inline qutebrowser/browser/hints.py]: I don't think hard-coding 15 here is a good idea.

We can get the current height of the label with

`self.height()`

however, at this point, the label dosen't seem to have been scaled to it's contents yet, and seems to have the default size.

One option here is to call `self.adjustSize()` to size the element before calling the height function. Note that there will be no text at this point, so the hint's width will be incorrect, but I think the height should be right. When calling `adjustSize` before `height`, I get a value of 16 on my computer (which may vary depending on the fonts used, for example).

File: qutebrowser/browser/webelem.py [hunk 2]
```diff
@@ -319,6 +319,11 @@ def _mouse_pos(self):
         else:
             rect.setHeight(rect.width())
         pos = rect.center()
+        if rect.y()<0 and (rect.y()+rect.height()-15>0):
```
Comment (jgkamat) [inline qutebrowser/browser/webelem.py]: In this case, why not just click the center of the 'visible' portion of the element (so we don't have to hardcode '15')? This was a bit hard to explain so I drew an image.

untitled

So, I think you may be able to do `rect.setY(max(0, rect.y()))`.

This will probably interact weirdly with the 'height/width' normalization on line 317. I think that putting something like this before that normalization would work, but I might have missed a corner case.

**Comment by jgkamat** [comment]
Oh, also it looked like some of the [tests are failing](https://travis-ci.org/qutebrowser/qutebrowser/builds/364678680) (angular buttons, and hints inside iframes). I can try to take a closer look at those later.

There are style tests as well (pylint and flake8).

**Comment by wentaozou1** [comment]
OK, I will try my best to fix it, but I am not 100% sure that I will solve it.

**Comment by jgkamat** [comment]
No problem, I'm happy to help you out if you get stuck along the way! :)

We're also in the `#qutebrowser` [channel on freenode](https://github.com/qutebrowser/qutebrowser#getting-help) if you would like to ask questions in semi-realtime, but I'm happy to answer questions here as well!

File: qutebrowser/browser/webelem.py (hunk 2 above)
Comment (wentaozou1) [reply qutebrowser/browser/webelem.py]: I have thought about using the solution you mentioned, but some problem will occurred when there is overlaps. For example, in the duckduckgo image search, the menu bar is on top of the image, and sometimes the click point will be on the menu bar. 

I have come up with another solution. Because the hint will be showed on the image, we could set the click point to the the location of the hint, which will guarantee the click point is on the intended link.

Comment (jgkamat) [reply qutebrowser/browser/webelem.py]: Hmm, this won' always work either because:

1. If the bar is on the botom, we will always hit the bottom of the image (meaning the image will always be unclickable)
2. If the image is thinner in height than the link, we will miss the element completely. (This is why the tests are failing, I think, some buttons in the test suite are shorter than 15px).

The proper solution to this is to use js to click the element directly (instead of sending these clicks), but I think we can keep that on hold for another PR. Using the fake-click method, there will always be cases where we don't click the correct element, but using the 'middle of visible area' method we can at least guarantee we will always hit an unobscured element properly.

**Comment by The-Compiler** [comment]
@wentaozou1 any update on this? If you need help, just let us know :smile:

**Comment by The-Compiler** [comment]
I'm closing this for now as it's inactive - @wentaozou1 if you want to continue working on this or need help, please let us know!
